python/whylogs/core/declarative_schema.py

```diff
@@ -0,0 +1,173 @@
+import logging
+from dataclasses import dataclass, field
+from typing import Any, Dict, List, Optional
+
+from whylogs.core.datatypes import (
+    AnyType,
+    DataType,
+    Fractional,
+    Integral,
+    String,
+    TypeMapper,
+)
+from whylogs.core.metrics import StandardMetric
+from whylogs.core.metrics.metrics import Metric, MetricConfig
+from whylogs.core.resolvers import Resolver
+from whylogs.core.schema import ColumnSchema, DatasetSchema
+from whylogs.core.segmentation_partition import SegmentationPartition
+
+logger = logging.getLogger(__name__)
+
+
+"""
+The DeclarativeSchema allows one to customize the set of metrics
+tracked for each column in a data set. Pass its constructor a list
+of ResolverSpecs, which specify the column name or data type to
+match and the list of MetricSpecs to instantiate for matching columns.
+Each MetricSpec specifies the Metric class and MetricConfig to
+instantiate. Omit the MetricSpec::config to use the default MetricConfig.
+
+For example, DeclarativeResolver(resolvers=STANDARD_RESOLVER) implements
+the same schema as DatasetSchema(), i.e., using the default MetricConfig,
+StandardTypeMapper, StandardResolver, etc.
+"""
+
+
+@dataclass
+class MetricSpec:
+    """
+    Specify a Metric to instantiate.
+    """
+
+    metric: Any  # Should be a subclass of Metric, it should be the class, not an instance
+    config: Optional[MetricConfig] = None  # omit to use default MetricConfig
+
+
+@dataclass
+class ResolverSpec:
+    """
+    Specify the metrics to instantiate for matching columns. column_name
+    takes precedence over column_type. column_type should be a subclass
+    of DataType, i.e., AnyType, Frational, Integral, or String. Pass the
+    class, not an instance.
+    """
+
+    column_name: Optional[str] = None  # TODO: maybe make this a regex
+    column_type: Optional[Any] = None
+    metrics: List[MetricSpec] = field(default_factory=list)
+
+
+# whylabs expects COLUMN_METRICS to be present for every column.
+COLUMN_METRICS = [MetricSpec(StandardMetric.counts.value), MetricSpec(StandardMetric.types.value)]
+
+
+# STANDARD_RESOLVER matches the default DatasetSchema/StandardResolver behavior
+STANDARD_RESOLVER = [  # TODO: maybe move this to unit test?
```

Comment:
why not move it to the unit test now?

Reply:
I'm not sure how useful it is beyond documentation. `DatasetSchema()` already implements the default behavior :)  Maybe move them to the example notebook one that exists...